File: OfflineClient/app/src/main/java/de/jbamberger/offlineclient/di/ViewModelModule.java
package de.jbamberger.offlineclient.di;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;

import dagger.Binds;
import dagger.Module;
import dagger.multibindings.IntoMap;
import de.jbamberger.offlineclient.ui.jodel.feed.JodelFeedViewModel;
import de.jbamberger.offlineclient.ui.reddit.RedditViewModel;
import de.jbamberger.offlineclient.ui.viewutil.AppViewModelFactory;

@Module
abstract class ViewModelModule {

    @Binds
    @IntoMap
    @ViewModelKey(RedditViewModel.class)
    abstract ViewModel bindRedditViewModel(RedditViewModel viewModel);

    @Binds
    @IntoMap
    @ViewModelKey(JodelFeedViewModel.class)
    abstract ViewModel bindJodelFeedViewModel(JodelFeedViewModel userViewModel);

    @Binds
    abstract ViewModelProvider.Factory bindViewModelFactory(AppViewModelFactory factory);
}

File: OfflineClient/app/src/main/java/de/jbamberger/offlineclient/di/ViewModelModule.kt
package de.jbamberger.offlineclient.di

import android.arch.lifecycle.ViewModel
import android.arch.lifecycle.ViewModelProvider
import dagger.Binds
import dagger.Module
import dagger.multibindings.IntoMap
import de.jbamberger.offlineclient.ui.main.MainActivityViewModel
import de.jbamberger.offlineclient.ui.viewutil.AppViewModelFactory

@Module
internal abstract class ViewModelModule {

    @Binds
    @IntoMap
    @ViewModelKey(MainActivityViewModel::class)
    internal abstract fun bindMainActivityViewModel(viewModel: MainActivityViewModel): ViewModel

    @Binds
    internal abstract fun bindViewModelFactory(factory: AppViewModelFactory): ViewModelProvider.Factory
}
